app/javascript/src/LoginPage/loginForms/Login3scaleForm.jsx

```diff
@@ -33,31 +47,26 @@ class Login3scaleForm extends React.Component<Props, State> {
     }
   }
 
-  setIsValidUsername = () => {
-    this.setState({isValidUsername: this.state.username !== ''})
-  }
-
   handleTextInputUsername = (username: string) => {
-    this.setState({ username }, this.setIsValidUsername)
-  }
-
-  setIsValidPassword = () => {
-    this.setState({isValidPassword: this.state.password !== ''})
+    const usernameError = validate.single(username, {presence: true, length: {minimum: 1}})
+    const isValidUsername = !usernameError
+    this.setState({ username, isValidUsername })
   }
 
   handleTextInputPassword = (password: string) => {
-    this.setState({ password }, this.setIsValidPassword)
+    const passwordError = validate.single(password, {presence: true, length: {minimum: 1}})
+    const isValidPassword = !passwordError
+    this.setState({ password, isValidPassword })
   }
 
   validateForm = (event: SyntheticEvent<HTMLInputElement>) => {
-    this.setIsValidUsername()
-    this.setIsValidPassword()
-    const isFormDisabled = this.state.username === '' ||
-      this.state.password === '' ||
-      this.state.isValidUsername === false ||
-      this.state.isValidPassword === false
-    if (isFormDisabled) {
+    const errors = validate(event.currentTarget.form, constraints)
+    if (errors) {
       event.preventDefault()
+      //$FlowFixMe: Needed due to a flow issue with Object values/keys: https://github.com/facebook/flow/issues/2221
+      for (const errorId in errors) {
+        this.setState({[namesToStateKeys[errorId]]: false})
```

Comment:
This might not be needed if you disable de form submit by default. You could also define your state like : 
```
{
username: '',
password: '',
validation: {
  username: undefined,
  password: undefined,
}
}
```
you'll be avoiding the declaration of https://github.com/3scale/porta/pull/1005/commits/36671e33b10614952825ac106c1e26e43af93773#diff-119386a465d8704baf0fcb20a45ed705R34

Then the check you can do is an `Array.every` iterating the validation keys and checking the values are `true`, then you avoid defining an extra step for checking the constrains.

What do you think?

Reply:
I think having validations in a object is 👌

Reply:
disable by default through javascript, right? (otherwise no js will have no way to enable the form, i know hypothetical)

Reply:
@thomasmaas are you OK with disabling the form by default?
Personally, I don't like it, I prefer disabling after the user has clicked on submit, because I think is good to give the user time to fill the form

Reply:
IMHO having validation on submit is the most annoying thing in a form 🤣 I usually want to know what's wrong the earliest possible so that I don't waste time clicking submit.